pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,287 @@
+func setGPIONodePhysAddrHelper(currNodePath string, physAddress uint64, numCAddrCells uint32) (uint64, error) {
+
+	if currNodePath == dtBaseNodePath { // Base Case: We are at the root of the device tree.
+		return physAddress, nil
+	}
+
+	// Normal Case: We are not at the root of the device tree. We must continue mapping our child addr (from the previous call) to this parent's addr space.
+	parentNodePath := filepath.Dir(currNodePath)
+	numPAddrCells, numAddrSpaceCells, err := getNumAddrSizeCellsInfo(parentNodePath)
+	if err != nil {
+		return INVALID_ADDR, err
+	}
+
+	var addrRanges []rangeInfo
+	if physAddress == INVALID_ADDR { // Case 1: We are the Child Node. No addr has been set. Read the reg file to get the physical address within our parents space.
+		physAddress, err = getRegAddr(currNodePath, numPAddrCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+	} else { // Case 2: We use the ranges property to continue mapping a child addr into our parent addr space.
+		addrRanges, err = getRangesAddrInfo(currNodePath, numCAddrCells, numPAddrCells, numAddrSpaceCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+		// getRangesAddrInfo returns a list of all possible child address ranges our physical address can fall into. We must see which range to use, so that we can map our physical address into the correct parent address range.
+		for _, addrRange := range addrRanges {
+
+			if addrRange.childAddr <= physAddress && physAddress <= addrRange.childAddr+addrRange.addrSpaceSize {
+				physAddress -= addrRange.childAddr  // get the offset beween the address and child base address
+				physAddress += addrRange.parentAddr // now address has been mapped into parent space.
+				break
+			}
+		}
+	}
+
+	numCAddrCells = numPAddrCells
+	currNodePath = parentNodePath
+	return setGPIONodePhysAddrHelper(currNodePath, physAddress, numCAddrCells)
+}
+
+// Creates a virtual Page to access/manipulate memory related to gpiochip data
+func (b *pinctrlpi5) createGPIOVPage(memPath string) error {
+	var err error
+
+	// Open the 'file' you are trying to map. Note: /dev/gpiomem0 is an alias, so it's file length returns 0
+	fileFlags := os.O_RDWR | os.O_SYNC
+	b.memFile, err = os.OpenFile("/dev/gpiomem0", fileFlags, 0666) // 0666 is an octal representation of: file is readable / writeable by anyone
+	if err != nil {
+		return fmt.Errorf("failed to open %s: %w\n", memPath, err)
+	}
+
+	pageSize := uint64(syscall.Getpagesize())
+	pageOffset := b.physAddr & (pageSize - 1) // difference between base address of the page and the address we're actually tring to access
+	//pageStartAddr := int64(b.physAddr - pageOffset) // base address of the page. needed when we can't open dev/gpiomem0
+	lenMapping := int(pageOffset) + int(b.chipSize) // total amount of memory needed to be mapped. + pageoffset is because we're starting from the base address of the page, not our actual physical address
```

Comment:
I thought we determined we don't need to use the pageOffset here?

Reply:
I thought only the offset needed to be aligned with the page not the length?

Reply:
If you offset = 0 and you're data is actually located at the start of the page, then you don't need the length. But if it the offset is a nonzero value then your data starts from somewhere in the middle of the page, and you have to account for that by extending the length of the mapping, because the mapping always starts from the beginning of a page regardless of whether or not the data you want starts at the beginning. You need the sum of the length + the offset to ensure that you map enough pages into VM in the case where your data is located across multiple pages; the OS manages this extra mapping internally.

In the case of accessing gpiomem0, offset = 0 so we don't actually need it but we would in any other case where offset is not aligned with the beginning of apage.

Reply:
If the offset needs to be a multiple of the page size then it will always start an the beginning of a page and not in the middle no?

Reply:
that makes sense! can you put a short comment in the code explaining this.

Reply:
I tested this and realized I have a gap in my understanding. The OS will account for the total length of the mapped space != multiple of page blocks and always map full blocks, but you do need the adjusted length value. Otherwise the OS will never map the second portion of the data lying on the second page.

A better way to think about this conceptually is that the OS manages reads and writes at a block level, not a byte level. So you're mapping in multiples of blocks - this is why our start address must be aligned with the beginning of a block, and also why the OS maps in full block increments. 

As soon as one part of a block changes from a write, the OS will treat the entire block as being updated. 

read this for further info: https://stackoverflow.com/questions/47604431/why-we-can-mmap-to-a-file-but-exceed-the-file-size

**Back to the earlier question:**

To make this more concrete, here's an example of an edge case that will fail without the lenMapping variable:

Imagine 2 blocks of size 100. You want to read 75 bytes of data, starting from address 50 (aka block1[50]). 

You want to read block1[50 - 99], and block2[0 - 25]

Your offset must be aligned with the beginning of block1, (block1[0].) If you don't include the difference between the start of the page and start of your data,  you will only get block1[50] - block1[99]. This is because your len = 75, so the OS only mapped 1 page. It won't map the second page, because the len was under 100 bytes. The rest of the bytes you wanted to read will return as 0, or error. 

However if you included the difference between the start of your data & the start of the page (aka 50), the len = 125. Then the OS would map 2 pages, and now you can read data[50 - 125] correctly, which spans over block1 and block2.

If this doesn't maek sense I can diagram it out on the whiteboard. Also, a better name for PageOffset is dataStartingAddr.

Reply:
that makes a lot of sense thank you for the detailed explanation! I guess im just not sure we need this if we are only going to be mapping /dev/gpiomem

Reply:
is it worth leaving in there for future reference? I can get rid of it but in the future data may not align with the page address, and if someone uses this code they might want to know that? If anyone tried to use this method in a different context  it could fail

Reply:
yeah def agree we should keep for future reference, I dont like the idea of having code we're not using in the repo so can you comment out the two lines of code with an explanation similar to above?